overlord/snapstate/snapstate.go

```diff
@@ -2102,7 +2105,32 @@ func splitRefresh(essential, nonEssential []minimalInstallInfo, updateFunc func(
 			}
 
 			appStartTask.WaitFor(baseEndTask)
+			crossSetDependency = true
+		}
+	}
+
+	// essential snaps don't use epochs at the moment so we only need to consider
+	// re-refreshes for the non-essential refreshes
+	if len(nonEssentialTss.Refresh) > 0 && !flags.NoReRefresh {
+		rerefreshSetup := reRefreshSetup{
+			UserID: userID,
+			Flags:  flags,
+		}
+
+		// if there are no cross-set dependencies, the rerefresh can be done
+		// before the reboot. If some app does need to wait for the reboot, the
+		// rerefresh check needs to wait for it so we do it at the end as usual
+		if !crossSetDependency {
+			for _, ts := range nonEssentialTss.Refresh {
+				for _, t := range ts.Tasks() {
+					rerefreshSetup.TaskIDs = append(rerefreshSetup.TaskIDs, t.ID())
+				}
+			}
 		}
+
+		rerefresh := st.NewTask("check-rerefresh", reRefreshSummary(nonEssentialUpdated, flags))
```

Comment:
How painful/feasible would it be to create a check-rerefresh task for snaps that can be rerefreshed before a reboot, and then another one for snaps that need a reboot?

Then a singular cross-set dependency wouldn't prevent all of the snaps from being re-refreshed. Although I'm not sure how common that situation might be, so the complexity (and test updates) might not be worth the effort.

Reply:
Actually I was thinking that a refresh might trigger the installation of a default provider which complicates this a bit. It's also a problem without considering rerefreshes because an app might cause a default provider to be added to the change and that default provider may depend on the model base and right now it would run in parallel with it. I'll fix that now and come back to the improvement you mentioned later since its impact is lower (few snaps use epochs AFAIK and at worst we check it a bit later). Thanks for the suggestion